docs/proposals/scheduling/20230811-node-allocatable-cpu-amplification.md

```diff
@@ -0,0 +1,103 @@
+  * [Summary](#summary)
+  * [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals/Future Work](#non-goalsfuture-work)
+  * [Proposal](#proposal)
+    * [User Stories](#user-stories)
+    * [Architecture](#architecture)
+    * [Workflow](#workflow)
+    * [API](#api)
+    * [Notes](#notes)
+      * [Container CPU Utilization Monitoring](#container-cpu-utilization-monitoring)
+    * [Constraints](#constraints)
+    * [Risks and Mitigations](#risks-and-mitigations)
+<!-- TOC -->
+
+## Summary
+
+Introduce a mechanism to amplify the node allocatable CPU, which enables the node to meet higher pod CPU requests. This feature can then be utilized by other higher-level ones such as CPU normalization.
+
+## Motivation
+
+There are cases where users would like to amplify the allocatable CPU of a node, for example:
+
+- The node is equipped with a new-gen CPU, which is way faster than those on other nodes. In this case, users would like to amplify the allocatable CPU of the node in order to make most of the fast CPU of the node as well as to make containers with the same CPU requests gain consistent computing power on different nodes.
+- The node's allocatable CPU is nearly full but the actual CPU usage remains at an extremely low level. In this case, users would like to amplify the allocatable CPU of the node in order to improve the CPU utilization of the node as well as to increase the amount of schedulable pods of the cluster.
+
+### Goals
+
+- Introduce a mechanism to amplify the node allocatable CPU for CPUShare pods at runtime.
+- The amplification of the node allocatable CPU does not affect the CPU resource for CPUSet pods.
+
+### Non-Goals/Future Work
+
+- Amplify the node allocatable CPU for CPUSet pods.
+
+## Proposal
+
+### User Stories
+
+This has been covered in [motivation](#motivation).
+
+### Architecture
+
+![image](/docs/images/node-allocatable-cpu-amplification-architecture.svg)
+
+### Workflow
+
+On k8s side, user (end user or other system) labels the Node to specify its allocatable CPU amplification ratio.
+
+On koord-manager side, the Node mutating webhook intercepts the Node status updating request from kubelet and modifies its allocatable CPU according to the amplification ratio. It also records the raw allocatable CPU in annotations of the Node.
+
+On koordlet side:
+
+- A runtime hook intercepts the CPU resource allocating request from kubelet and allocates actual underlying CPU resource for CPUShare Pods according to the amplification ratio. It also keeps reconciling CPU resource for all CPUShare Pods on the Node to make sure the amplification is applied to any existing CPUShare Pod.
+- The node info metrics advisor amplifies the allocatable CPU in each NUMA node according to the amplification ratio, so that the allocatable CPU in `zones` of `NodeResourceTopology` of the Node would also be amplified. This can make sure the NUMA topology scheduling of CPUShare Pods working properly with amplified allocatable CPU.
```

Comment:
But we do need to amplify the allocatable CPU in real `NodeResourceTopology` resource here (not only in koordlet cache) just like the Nodes' `allocatable`, because the scheduler also needs this information to do NUMA topology scheduling for CPUShare Pods with amplified allocatable CPU. So we can only do this in metrics advisor I think.

Reply:
`amplify the allocatable CPU in real NodeResourceTopology resource` to let the scheduler know is also the states informer's work. The koordlet should be able to know the actual state of the node, so as to reconcile the cgroups and other sys files correctly, where the metrics advisor and the metric cache are the data sources.

Reply:
I see, changed.